Parameter Improvements

afscrome opened the pull request:
## Description

- Fixed the `ProcessResourceWithoutLifetime` method so it actually got called for resources without lifetime.  Previously it hooked into `ResourceEndpointsAllocatedEvent` which was only ever fired for Container and Executable resources.   This in turn properly hooks up Values to show up in the Dashboard.
- Fire the `ConnectionStringAvailable` event for `ConnectionStringResource` so that Connection String resoruces show the connection string exactly like other `IResourceWithConnectionString` that have lifetimes
- Introduced an `Active` state to be used for resources without a lifetime.
- Made `Value` a well known property of Parameter resources so it shows up on the dashboard without having to hit "Show more"
- Introduced a dedicated icons for Parameter & connection string resources
   - `Settings` as opposed to the default `SettingsCogMultiple` for parameter resources
   - `PlugConnectedSettings` for connection strings

Non secret Param
![image](https://github.com/user-attachments/assets/cbcce2c7-3d86-4958-962c-53d3a825ccf4)

Secret Param
![image](https://github.com/user-attachments/assets/11259d22-9d1b-4be3-bfcc-e410ef4496ec)

Erroring Param
![image](https://github.com/user-attachments/assets/40f614d8-db56-422f-a75c-8ad6b44b0322)

Connection String
![image](https://github.com/user-attachments/assets/61754469-8749-4ecc-96e7-3f3d20b1a741)


Fixes @maddymontaquila's complaints on AspiriFriday stream.

## Checklist

- Is this feature complete?
  - [X] Yes. Ready to ship.
  - [ ] No. Follow-up changes expected.
- Are you including unit tests for the changes and scenario tests if relevant?
  - [ ] Yes
  - [X] No - changes are all UI based
- Did you add public API?
  - [ ] Yes
    - If yes, did you have an API Review for it?
      - [ ] Yes
      - [ ] No
    - Did you add `<remarks />` and `<code />` elements on your triple slash comments?
      - [ ] Yes
      - [ ] No
  - [X] No
- Does the change make any security assumptions or guarantees?
  - [ ] Yes
    - If yes, have you done a threat model and had a security review?
      - [ ] Yes
      - [ ] No
  - [X] No
- Does the change require an update in our Aspire docs?
  - [ ] Yes
    - Link to aspire-docs issue (consider using one of the following templates):
      - [New (or update) `doc-idea` template](https://github.com/dotnet/docs-aspire/issues/new?template=02-docs-request.yml)
      - [New `breaking-change` template](https://github.com/dotnet/docs-aspire/issues/new?template=04-breaking-change.yml)
      - [New `diagnostic` template](https://github.com/dotnet/docs-aspire/issues/new?template=06-diagnostic-addition.yml)
  - [X] No

afscrome committed:
Fixed the `ProcessResourceWithoutLifetime` method so it actually got called for resources without lifetime.
 Previously it hooked into `ResourceEndpointsAllocatedEvent` which was only ever fired for Container and Executable resources.
This in turn properly hooks up Values to show up in the Dashboard.

Introduced an `Active` state to be used for resources without a lifetime.

Made `Value` a well known property of value resources so it shows up on the dashboard without having to hit "Show more"

Introduced a dedicated icon (`Settings` as opposed to the default `SettingsCogMultiple`) for parameter resources

davidfowl commented on src/Aspire.Hosting/ParameterResourceBuilderExtensions.cs:
```diff
@@ -165,12 +166,20 @@ internal static IResourceBuilder<T> AddParameter<T>(this IDistributedApplication
             State = KnownResourceStates.Running
         };
 
+        builder.Eventing.Subscribe<AfterResourcesCreatedEvent>(async (evt, ct) =>
```
Not sure we want to use this one,

afscrome replied:
Ah, `ProcessResourceWithoutLifetime` didn't show up when I started writing my comment, I'll take a look at that.

davidfowl commented on src/Aspire.Hosting/Orchestrator/ApplicationOrchestrator.cs:
```diff
@@ -265,10 +266,19 @@ private async Task ProcessResourceUrlCallbacks(IResource resource, CancellationT
+            return;
+        }
+
+        await ProcessResourceWithoutLifetime(@event.Resource, cancellationToken).ConfigureAwait(false);
```
We may have to defer resources that reference endpoints to be resolved later:

https://github.com/dotnet/aspire/blob/5f415f762891ee253f42a5940c0cce948ac3b8fd/playground/ParameterEndToEnd/ParameterEndToEnd.AppHost/Program.cs#L24

davidfowl commented on src/Aspire.Hosting/Orchestrator/ApplicationOrchestrator.cs:
```diff
@@ -117,6 +117,11 @@ private async Task OnEndpointsAllocated(OnEndpointsAllocatedContext context)
             await lifecycleHook.AfterEndpointsAllocatedAsync(_model, context.CancellationToken).ConfigureAwait(false);
         }
+
+        foreach (var resource in _model.Resources.OfType<IResourceWithoutLifetime>())
```
Undo this change and put this back where it was. The layering was better.

afscrome replied:
Think I have a better solution now, although it does rely on treating `ParameterResource` and `ConnectionStringExpression` specifically rather than generically handle all `IResourceWithoutLifetimes` that are also `IManifestExpressionProvider`.  Think it's good, will do some more validation tomorrow.

This was supposed to be a one-liner when I first started 😅.

afscrome committed:
Improve Connection String Resource handling

- `ConnectionStringResource` now raises `ConnectionStringAvailable` event, so it gets the same `ConnectionString` property emitted as all other `IResourceWithConnectionString` resources that do have lifetimes.
- Reworked `ConnectionStringResource` initial state to start in a `Waiting` state, and then transition to `Active` once the connection string is evaluated
- Reworked `ParameterResource` to use same state approach as Connection string - start in `Waiting` state

afscrome committed:
Add icon for connection strings

afscrome committed:
Unhide Parameter resources if they error

afscrome commented on src/Aspire.Hosting/Orchestrator/ApplicationOrchestrator.cs:
```diff
@@ -265,44 +266,103 @@ private async Task ProcessResourceUrlCallbacks(IResource resource, CancellationT
+                    {
+                        State = new("Value missing", KnownResourceStateStyles.Error),
+                        Properties = s.Properties.SetResourceProperty("Value", ex.Message),
+                        IsHidden = false
```
Doesn't fully work due to #10121.

afscrome committed:
Add a few more scenarios  to ParameterEndToEnd test

afscrome committed:
Use `KnownProperties` when setting parameter value.

davidfowl commented:
@afscrome some conflicts!

afscrome committed:
Merge branch 'main' into improve-parameters
